### PR TITLE
Add pomodoro timer

### PR DESCRIPTION
**Pull request by blue2wo**
## Goal
Create the Pomodoro Timer component and the integration tests for this component.

## How to test
1. Run `yarn test` making sure all tests pass. There will be React errors stating that the <svg> children elements such as <circle> or <g> can't be recognized by the browser. Ignore this for now as the tests still pass and the app works correctly, it's only an issue within the test simulated browser.
2. If all tests pass, check the timer visuals, comparing against the designs. Note: Responsiveness or changing timer settings has not been implemented yet.   3. Check timer functionality. Timer should countdown from 25:00 to 00:00 during focus time then 05:00 to 00:00, continuing in this loop. You will have to manually start each countdown ends.
Functionality to check:
- [ ] Focus time ends and switches to break time
- [ ] Break time ends and switches to focus time
- [ ] Pause the timer
- [ ] Start the timer

## Screenshots
image
image
image

### CODE REVIEW

File: __tests__/index.test.tsx
Comment (Evomatic) [inline __tests__/index.test.tsx:1]: Note: `index.tests.tsx` was pushed by accident. make sure to remove it.

File: __tests__/functions/functions.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  it("should return 0 for negative values", () => {
+    // When given a negative numerical value
+    expect(cleanPercentage(-15)).toBe(0);
+
+    // When given a negative alphabetic value
+    expect(cleanPercentage("-15")).toBe(0);
+  })
+
+  it("should return 100 for values greater than 100", () => {
+    // When given a negative numerical value
```
Comment (Evomatic) [inline __tests__/functions/functions.ts:21]: I think the comment is misleading here. How about saying something along the lines of, "When given a numeric value greater than 100"

File: __tests__/functions/functions.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    expect(cleanPercentage(110)).toBe(100);
+
+    // When given a negative alphabetic value
```
Comment (Evomatic) [inline __tests__/functions/functions.ts:24]: Here as well, lets fix this comment!

File: __tests__/functions/functions.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+    expect(cleanPercentage("110")).toBe(100);
+  })
+
+  it("should return 0 for NaN values", () => {
+    expect(cleanPercentage("abc")).toBe(0);
+  })
+
+  it("testing edge cases", () => {
```
Comment (Evomatic) [inline __tests__/functions/functions.ts:32]: For the edge cases, I think they deserve their own describe block. We can nest describes to better organize and isolate exactly what we want to test. 

```TS

describe("Test edge cases", () => {
    it("Return 0 when given 0", () => {

      expect(cleanPercentage(0)).toBe(0);
    })

    it("Return 100 when given 100", () => {

       expect(cleanPercentage(100)).toBe(100);
    })

    it("Return 100 when given a super duper pumper to the maxMaxMax number", () => {

      expect(cleanPercentage(Number.MAX_VALUE)).toBe(100);
   })

  })
```

File: app/src/components/circular-progress.tsx [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  const circ = 2 * Math.PI * r;
+  const strokePercentage = ((100 - +percentage) * circ) / 100; // where stroke will start, e.g. from 15% to 100%.
+  return (
+    <circle
+      data-testid={`progress-circle-${colour}-test`}
+      r={r}
+      cx={-5}
+      // cy={-105} // Can remove this and inclde in transform translate attribute
+      fill="transparent"
+      stroke={strokePercentage !== circ ? colour : ''} // remove colour as 0% sets full circumference
```
Comment (Evomatic) [inline app/src/components/circular-progress.tsx:29]: Lets remove too much logic happening inside the property and simplify it by adding to a variable first then passing the variable to the prop.

```ts
const stroke = strokePercentage !== circ ? colour : ''
```

File: app/src/components/circular-progress.tsx [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+      strokeWidth={15}
+      strokeDasharray={circ}
+      strokeDashoffset={percentage ? strokePercentage : 0}
```
Comment (Evomatic) [inline app/src/components/circular-progress.tsx:32]: Here as well, lets not pass the logic in the prop

File: app/src/components/circular-progress.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+}: {
+  minutes: string;
+  seconds: string;
+}): ReactElement => {
+  return (
+    // https://developer.mozilla.org/en-US/docs/Web/SVG/Element/text
+    // 21.66 is the width of the semicolon html element
+    <g data-testid="render-timer-text-test">
+      <StringToTextElement xvalue={27} text={minutes} />
+      <text className="fill-on-dark-background text-8xl" x={205 - 21.66 / 2} y={205 + 21.66}>
```
Comment (Evomatic) [inline app/src/components/circular-progress.tsx:76]: The `x` and `y` props, lets pass the logic to variables first!

File: app/src/components/pomodoro-timer.tsx [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        setTimeRemaining(1500);
+      }
+      setIsTimerActive(false);
+    }
+  }, [timeRemaining, isFocusTime]);
+
+  return (
+    <div className="h-[410px] w-[410px]">
+      <div className="relative flex flex-col items-center justify-center rounded-full bg-primary-dark">
+        <CircularProgress percentage={isFocusTime ? focusTimePercentage : breakTimePercentage} colour="#f87070" minutes={minutes} seconds={seconds} />
```
Comment (Evomatic) [inline app/src/components/pomodoro-timer.tsx:60]: For the `percentage` prop lets assign to a variable and pass it.

**Review by Evomatic — CHANGES_REQUESTED** [review]
Thinks are looking good! I left a few comments on ways to improve the readability of the code. Also I noticed some issues with the timer, but maybe we can revisit them in a separate PR.

- When under 10 minutes the timer format is incorrect. Too much space for minutes. Should be `01:24` not `1 :24`

**Comment by Evomatic** [comment]
don't forget to format the files at the very end! 📝

**Comment by Evomatic** [comment]
@blue2wo I merged main branch after the addition to the layout and added the timer to the layout. Looks good on desktop! But the timer is clipped when mobile, due to no responsive design for the component. No problem though, we can fix this in separate PR along with some other fixes that need to happen.

Desktop:
image

Mobile: 375 x 667
image


LGTM! 🧑‍💻

**Review by Evomatic — APPROVED** [review]
LGTM! 🧑‍💻

**Review by Evomatic — APPROVED** [review]

**Review by Evomatic — APPROVED** [review]
